cirq/google/optimizers/convert_to_sqrt_iswap.py

```diff
@@ -0,0 +1,290 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import List, cast
+
+import numpy as np
+
+import cirq
+import sympy
+
+SQRT_ISWAP = cirq.ISWAP**-0.5
+
+
+# TODO: Combine this with the equivalent functions in google/gate_set.py
+# Or better yet, write a proper gate set so we don't need this in two places
+def _near_mod_n(e, t, n, atol=1e-8):
+    if isinstance(e, sympy.Basic):
+        return False
+    return abs((e - t + 1) % n - 1) <= atol
+
+
+def _near_mod_2pi(e, t, atol=1e-8):
+    return _near_mod_n(e, t, 2 * np.pi, atol=atol)
+
+
+class ConvertToSqrtIswapGates(cirq.PointOptimizer):
+    """Attempts to convert gates into ISWAP**-0.5 gates.
+
+    Since we have Z rotations and arbitrary XY rotations, we
+    can rely on cirq decomposition for one qubit gates and
+    need to only specify special decompositions for two qubit gates.
+
+    Currently natively specified gates are CZPowGate, ISwapPowGate,
+    and FSimGate.  This will also support gates that decompose into
+    the above gates.
+    """
+
+    def __init__(self, ignore_failures=False) -> None:
+        """
+        Args:
+            ignore_failures: If set, gates that fail to convert are forwarded
+                unchanged. If not set, conversion failures raise a TypeError.
+        """
+        super().__init__()
+        self.ignore_failures = ignore_failures
+
+    def _convert_one(self, op: cirq.Operation) -> cirq.OP_TREE:
+        """
+        Decomposer intercept:  Let cirq decompose one-qubit gates,
+        intercept on 2-qubit gates if they are known gates.
+        """
+        if isinstance(op, cirq.GlobalPhaseOperation):
+            return []
+
+        gate = cast(cirq.GateOperation, op).gate
+
+        if len(op.qubits) != 2:
+            return NotImplemented
+
+        if isinstance(gate, cirq.CZPowGate):
+            turns = cast(cirq.CZPowGate, gate).exponent
```

Comment:
I think mypy understands after an isinstance check that `gate` is an instance of `cirq.CZPowGate` so this cast should not be needed. Same thing for the other cases below. Also, consider calling the variable `exponent` instead of `turns`.

Reply:
Weird.  I totally thought mypy complained about this, but I guess not.